x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/Verifier.java

```diff
@@ -59,20 +63,19 @@
     }
 
     private static void checkForNonCollapsableSubselects(PhysicalPlan plan, List<Failure> failures) {
-        Holder<Boolean> hasLimit = new Holder<>(Boolean.FALSE);
-        Holder<List<Order>> orderBy = new Holder<>();
-        plan.forEachUp(p -> {
-            if (hasLimit.get() == false && p instanceof LimitExec) {
-                hasLimit.set(Boolean.TRUE);
-                return;
-            }
-            if (p instanceof OrderExec) {
-                if (hasLimit.get() && orderBy.get() != null && ((OrderExec) p).order().equals(orderBy.get()) == false) {
-                    failures.add(fail(p, "Cannot use ORDER BY on top of a subquery with ORDER BY and LIMIT"));
-                } else {
-                    orderBy.set(((OrderExec) p).order());
-                }
+        // use a comparator to ensure deterministic order of the error messages
+        Set<LimitExec> limits = new TreeSet<>(Comparator.comparing(l -> l.source().text()));
+
+        plan.forEachDown(p -> {
+            if (p instanceof OrderExec || p instanceof FilterExec || p instanceof PivotExec || p instanceof AggregateExec) {
```

Comment:
No need to do double iteration - the main forEach already traverses the tree so for each matching node doing another p.forEachDown is unnecessary.
Further more I don't think it's useful to the user to see all the instances where the same type of error occurs since a query with ORDER BY and a WHERE clause will report the error twice (once per entity).

I favor the initial approach since it's O(n) - traverse the tree once. (this could be optimized by ending the traversal on the first error however it's a rare enough case that we can ignore it).

Reply:
👍 for the single iteration.

Also reporting only one error shouldn't make too much of a difference but I was more thinking about the cases where there is more than one offending `Limit` as in `SELECT * FROM (SELECT * FROM (SELECT * FROM test_emp LIMIT 10) LIMIT 20) WHERE languages > 5`. Since this seems to be a rather far-fetched case, the simpler one-error implementation should be fine.